packages/blockchain/src/index.ts

```diff
@@ -737,55 +737,55 @@ export default class Blockchain implements BlockchainInterface {
    * @param name - Name of the state root head
    * @param onBlock - Function called on each block with params (block, reorg)
    * @param maxBlocks - How many blocks to run. By default, run all unprocessed blocks in the canonical chain.
+   * @returns number of blocks actually iterated
    */
-  async iterator(name: string, onBlock: OnBlock, maxBlocks?: number) {
+  async iterator(name: string, onBlock: OnBlock, maxBlocks?: number): Promise<number> {
     return this._iterator(name, onBlock, maxBlocks)
   }
 
   /**
    * @hidden
    */
-  private async _iterator(name: string, onBlock: OnBlock, maxBlocks?: number) {
-    await this.initAndLock<void>(async () => {
-      const blockHash = this._heads[name] || this._genesis
-      let lastBlock: Block | undefined
-
-      if (!blockHash) {
-        return
-      }
-
-      if (maxBlocks && maxBlocks < 0) {
-        throw 'If maxBlocks is provided, it has to be a non-negative number'
-      }
+  private async _iterator(name: string, onBlock: OnBlock, maxBlocks?: number): Promise<number> {
+    return await this.initAndLock<number>(
+      async (): Promise<number> => {
+        const headHash = this._heads[name] || this._genesis
+        let lastBlock: Block | undefined
```

Comment:
I've also renamed some variable names for better clarity, not sure why this is show with such a large gap here, might be a bit confusing, but code flow remains the same here.